### DISCUSSION TITLE
Swarm data logging - How to access lambda function for each crazyflie

### DISCUSSION THREAD

**Discussion by Vrinda73** [bitcraze/discussions#794] [Q&A] [answered]
Hello !
I am currently playing with 4 crazyflies and I need to know the angular rates of each crazyflie to control them in a swarm. I was successful in writing a lambda function (inline with discussions on this page regarding logging). However I need to use them in a for loop for each crazyflie. I tried for few days to solve this problem but unsuccessful. Please can anyone help ? :( 
Thank you! My code is as follows

```
def log_stab_callback(uri, timestamp, data, log_conf):
        
        x = float(data['gyro.x'])
        y = float(data['gyro.y'])

def simple_log_async(scf):
    lg_vars = {
        'gyro.x': 'float',
        'gyro.y': 'float',
        
    }

    lg_stab = LogConfig(name='Gyro', period_in_ms=10)
    for key in lg_vars:
        lg_stab.add_variable(key, lg_vars[key])
    
    cf = scf.cf 
    cf.log.add_config(lg_stab)
    lg_stab.data_received_cb.add_callback(lambda t, d, l: log_stab_callback(cf.link_uri, t, d, l))
    lg_stab.start()
    
if __name__ == '__main__':
   
    cflib.crtp.init_drivers(enable_debug_driver=False) # initialize drivers
    factory = CachedCfFactory(rw_cache='./cache')

    with Swarm(uris, factory=factory) as swarm:
    		swarm.parallel(simple_log_async)
    		
    		while True:
                        x = simple_log_async("URI1")
                        print(x)

```

**Comment by krichardsson** [comment]
```python
import time
import cflib.crtp
from cflib.crazyflie.log import LogConfig
from cflib.crazyflie.swarm import CachedCfFactory
from cflib.crazyflie.swarm import Swarm
from cflib.crazyflie.syncLogger import SyncLogger

URI1 = 'radio://0/30/2M/E7E7E7E7E7'
URI2 = 'radio://0/30/2M/E7E7E7E7E8'

uris = {
    URI1,
    URI2,
}

def log_sync(scf):
    lg_vars = {
        'gyro.x': 'float',
        'gyro.y': 'float',

    }

    lg_stab = LogConfig(name='Gyro', period_in_ms=10)
    for key in lg_vars:
        lg_stab.add_variable(key, lg_vars[key])

    with SyncLogger(scf, lg_stab) as logger:
        endTime = time.time() + 3

        for log_entry in logger:
            uri = scf.cf.link_uri
            timestamp = log_entry[0]
            data = log_entry[1]

            print(f'{uri}, {timestamp}, {data}')
            # Do your stuff here

            if time.time() > endTime:
                break


if __name__ == '__main__':

    cflib.crtp.init_drivers(enable_debug_driver=False) # initialize drivers
    factory = CachedCfFactory(rw_cache='./cache')

    with Swarm(uris, factory=factory) as swarm:
        swarm.parallel(log_sync)
```

**Reply by Vrinda73** [reply]
Thank you very much Kristoffer! I was wondering how I can give attitude setpoints to each crazyflie in a swarm in the place you mentioned in the code. roll_setpoint1(for URI1) = function(roll rate of URI1). Once I have setpoints of all drones, how can I command all CFs via run_sequence in the place shown by you in the code. Again thank you so much. I am working alone on this project and your help is greatly valued.

**Comment by krichardsson** [comment, marked as answer]
If you want to pass static parameters to the function used in the swarm, [you can do like in this example](https://github.com/bitcraze/crazyflie-lib-python/blob/1f2c36105b87e1cf329113d61a2ca1dbdb7143c7/examples/autonomy/circling_square_demo.py#L219-L234)

If you want a more dynamic behavior you can use a function called from function used in the swarm (`log_sync` in the example above), like you suggest.

To send a setpoint to the Crazyflie you can use one of the setpoint methods in the [commander](https://github.com/bitcraze/crazyflie-lib-python/blob/master/cflib/crazyflie/commander.py) object of the Crazyflie, something like:
```python 
scf.cf.commander.send_setpoint(roll, pitch, yawrate, thrust)
```

**Reply by Vrinda73** [reply]
Thank you Kristoffer! It worked!
